gaps/gap-24_agreement_permissions_management/gap-24_agreement_permissions_management.md

```diff
@@ -0,0 +1,188 @@
+|-|-|
+| req:getAgreement | |
+| req:getAgreementPermissions | |
+| req:collectAgreementEvents | |
+| req:getActivityState | |
+| req:getActivityUsage | |
+| req:getRunningCommand | |
+| req:getExecBatchResults | |
+| req:getDebitNotes | |
+| req:getDebitNote | |
+| req:getPaymentsForDebitNote | |
+| req:getDebitNoteEvents | |
+| req:getInvoices | |
+| req:getInvoice | |
+| req:getPaymentsForInvoice | |
+| req:getInvoiceEvents | |
+| req:getPayments | |
+| req:terminateAgreement | |
+| req:setAgreementPermissions | See: [Transitive permissions](#transitive-permissions) below |
+| req:createActivity | |
+| req:exec | |
+| req:callEncrypted | |
+| req:destroyActivity | |
+| req:acceptDebitNote | See: [Accept Invoice and Debit Note logic](#accept-invoice-and-debit-note-logic) below |
+| req:rejectDebitNote | |
+| req:acceptInvoice | See: [Accept Invoice and Debit Note logic](#accept-invoice-and-debit-note-logic) below |
+| req:rejectInvoice | |
+
+If a permission to a specific REST API action on a given Agreement is granted to a Grantee, this Grantee is allowed to make respective REST API call referring to the Agreement (or related entity). 
+
+**Note:** this GAP deliberately leaves Provider-side action permissions out of scope.
+
+### Access Control List
+
+An Agreement shall have an *Access Control List* (ACL) associated with it, to indicate all rights granted to any non-Owner node/ID.
+
+An indicative example of an Agreement's ACL is shown below:
+
+```
+agreementACL = {
+    aclEntries: [
+        {
+            "grantee": "0x7735df0c0bbb3ca65c55d61eadded653573e0152",
+            "permissions" : [
+                "req:createActivity",
+                "req:exec",
+                "req:destroyActivity"
+            ]
+        },
+        ...
+        {
+            "grantee": "0x0053b2c00006d1d8b75bb7dfa068990a19e4c32e",
+            "permissions" : [
+                "req:terminateAgreement",
+                "req:acceptInvoice",
+                "req:rejectInvoice"
+            ]
+
+        }
+    ]
+}
+```
+
+### MarketAPI: SetAgreementPermissions
+
+A new method shall be added in Market API to enable setting of an Agreement's ACL.
+
+**Who:**
+- Requestor
+
+**Input:**
+- agreementId
+- new ACL content
+
+**Output:**
+- OK
+- Error 
+
+### MarketAPI: GetAgreementPermissions
+
+A new method shall be added in Market API to enable retrieving an Agreement's ACL.
+
+**Who:**
+- Requestor
+
+**Input:**
+- agreementId
+
+**Output:**
+- OK + Agreement's current ACL 
+- Error 
+
+### Accept Invoice and Debit Note logic
+
+**Note** that one of the parameters required for `AcceptDebitNote`/`AcceptInvoice` Payment API calls is an `allocationId`. The ACL allows for delegation of rights to an Agreement and associated entities, but there is no equivalent rights management for funds managed/owned by Agreement Owner. 
+
+Therefore a rights Grantee, who is willing to Accept an incoming Debit Note or Invoice must pass `allocationId` of their own Allocation.
```

Comment:
okay, in this simple model, it's really unclear how the payment process would work in case there is more than one grantee ... who pays? anyone? do we have a "race" to pay? what incentive would grantees have to be the ones to pay?

Reply:
- we abstract from incentives and intents to pay - we only allow for different requestors to pay.

- should we raise a "bug" to core team to ensure `AcceptDebitNote`/`AcceptInvoice` (as well as `Reject`) is race-resistant, ie. only a single accept on Invoice/DebitNote should end with 'OK', subsequent attempts should return info "invoice/debitnote already accepted". Must also ensure for this to be exercised on provider-side, as there must be a single-source-of truth and atomicity achieved for Accept operations. 
  - we ignore scenarios where a malicious provider may want to change the sequence of accept/reject messages (or hold/ignore incoming messages)

- how to solve the need to propagate Agreement/Activity related events over the "entitled" Requestors. We need to discuss this with the core team [ACTION - setup workshop with core team].

- "Rogue grantee" problem. What options do we have?
  1. Have a consensus mechanism over `setAgreementPermissions` 
    a. any 'setAgreementPermissions' permission amendment must be controlled by a forum of grantees
    b. any permission amendment must be individually controlled by a forum
    c. only change in set of grantees must be controlled by a forum

Reply:
"Rogue grantee" problem
- Option II doesn't solve a problem. We can have consensus on upper layer, but yagna must be consensus aware, otherwise one `grantee` can still Agreement despite consensus being against him.
    - But in this case grantees keeping consensus correctly could drop the Agreement and recreate it with new Node without giving permissions to malicious actor

Other options:
- (Initially) split upgrading permissions from downgrading. I would say that giving permissions to new Node is less harmful than stealing Agreement

Reply:
Mtg w/Witek on 14.02.:

1. On API "races" - this must simply be specified as a requirement, and included in test scenarios. NOTE: we should explictly list relevant Test Cases in the test suite. Also include "conflict races", ie. one requestor Accepts another Rejects...
2.  On entity and event propagation:
  a.   currently yagna (on network level) does not have QoS features, ie. it does not support "retry" on messages sent between sides. This means that if Requestor is offline, the Provider is unable to reliably send TerminateAgreement message (currently one side is unable to Terminate Agreement while the other side is offline). This seems to be a gap from the point of view of "Offline Requestor model"!!!
  b. it seems there is a gap in yagna core which currently makes 'attach/detach' impossible
  c. agreementId/nodeId propagation:
     1) either the network must notify the new grantee with agreementIds
     2) or we add REST API call to "force synchronization" with a specified node

Reply:
Update on 24.02.:
- Enhanced the GAP-24 with more detailed test case specification.

Reply:
7.03.2023:
- Ad ACL permissions, grantees, delegation - amend ACL to include various "credential sets" required (optional) where node identity is one of credential types, but also other credential types may be required (eg. a multi-sig artifact).
- Kamil to propose an illustrative example of ACL
- Ad. Agreement information propagation across network (required to forward the AgreementId to a new grantee)
  - Step 1: (suboptimal) each node knowing Agreement's ACL should periodically distribute it to affected parties (all grantees as listed in ACL) - can we extract all the affected parties straight.
  - Step 2: (optimal) A dedicated, generic GAP describing the "new node welcome notification" mechanism, as well as node "synchronization with network"
    - Ask Witek to draft brief input into the GAP - summarize the concept as discussed on the meeting

Reply:
Below is example for defining structure with separate collection of credential providing methods and separate collection of different permission sets. 
Because those entries are interconnected and they need to refer from one to each other, using custom names help significantly.

Specific key names and values are provisional and **are subject of change**.

Relation between `authTypes` and `permissionGroups` could be included in `permissionGroups` (as shown below) or could be moved to analogical structure in `authTypes`.

```json
{
	"version": 123,
	"authTypes": {
		"basic": {
			"type": "basic",
			"password": "..."
		},
		"token": {
			"type": "bearer",
			"token": "..."
		},
		"cert": {
			"type": "pem",
			"privateKey": "...",
		},
		"nodesOwned": {
			"type": "nodeId",
			"nodes": [
				"node_id_kj24",
				"node_id_9fd8",
				"node_id_gq68",
			]
		},
		"nodesNotOwned": {
			"type": "nodeId",
			"nodes": [
				"node_id_kj43",
				"node_id_ml23",
				"node_id_8s0d"
			]
		}
	},
	"permissionGroups": {
		"group1": {
			"authTypes": [
				"basic",
				"nodeOwned"
			],
			"permissions": [
				"req:createActivity",
				"req:exec",
				"req:destroyActivity"
			]
		},
		"group2": {
			"authTypes": [
				"token",
				"nodeNotOwned",
				"cert"
			],
			"permissions": [
				"req:terminateAgreement",
				"req:acceptInvoice",
				"req:rejectInvoice"
			]
		}
	}
}
``

Reply:
Incorporated the proposed ACL improvements in the GAP draft, with some tweaks (eg. not "authTypes", but rather "attestationTypes", etc.)
We need to discuss the potentially useful attestation types, as we must not include any secrets in the ACL itself!

Reply:
Notes after 14.03.:
- Continue on "generic attestations" - need to align on what do the "content-dependent" attestations include? The attestation artifact (eg. ACL list) must be: 
  - Canonicalized to allow reliable verification
  - Specified element of the request - how is this specified? 
    - API method does specify which element of it can be covered by attestation (eg. ACL content in `SetAgreementPermissions`)
    - ...any other option?